src/storage/src/source/reclock/compat.rs

```diff
@@ -311,86 +231,94 @@ impl RemapHandle for PersistHandle {
                 .await
                 .expect("local since is not beyond read handle's since")
             {
-                let binding = unpack_binding(update.expect("invalid row"));
+                let binding = update.expect("invalid row");
+                let binding = Self::FromTime::try_from(binding).expect("invalid binding");
+                assert!(
+                    binding.partition().is_some() || self.partitioned_source,
+                    "invalid binding stored in persist {:?}",
+                    binding
+                );
                 self.pending_batch.push((binding, ts, diff));
             }
+
+            // Partitionless collections don't store FromTime::minimum() values
```

Comment:
Is this comment correct? The migration should make sure the minimum element is always there which means that here we should never be synthesizing timestamps

Reply:
The minimum element of `Partitioned<PartitionId, MzOffset>` is `(Bottom, Top)`, which is expressed as `Datum::Range((,))` (i.e. the infinite range)`; that cannot/should not be serialized into the non-Kafka remap shards, whose schema is only one column of `Datum::UInt64`. However, the fact that we require the conversions into `OffsetAntichain` present domain-covering sets means that we have to synthesize timestamps, at least as long as we have `OffsetAntichain` floating around. Let me know if I've misunderstood something or something has changed while this has been in flight.

Reply:
Right so in this `RemapHandleReader` implementation there are two cases:
1. We are reading a partitioned source, in which case there is no synthesizing work to do and we can simply read a `(row, ts, diff)` update, convert into a `(Partitioned<..>, ts, diff)` update and pass it on
2. We are reading a non-partitioned source, in which case the only thing we need to synthesize are exactly two updates. `(Partitioned::with_range(None, Some(PartitionId::None), MzOffset::from(0)), min_ts, 1)` and `(Partitioned::with_range(Some(PartitionId::None), None, MzOffset::from(0)), min_ts, 1)`

So the only "special" logic I'd expect is this:

```rust
if partitioned_source && !self.minimum_synthesized {
    let mut min_ts = T::minimum();
    min_ts.advance_by(self.as_of.borrow());
    self.pending_batch.extend([
        (Partitioned::with_range(None, Some(PartitionId::None), MzOffset::from(0)), min_ts, 1),
        (Partitioned::with_range(Some(PartitionId::None), None, MzOffset::from(0)), min_ts, 1)
    ]);
    self.minimum_synthesized = true;
}
```

Why do we need the rest of the special handling?

Reply:
Every remap shard expects to be initialized to `(Partitioned<PartitionId, MzOffset>::minimum(), Timestamp::minimum(), 1)` by `ReclockOperator::new`. However, in the case of PG sources, we have to filter out that update because it generates a `Datum::Range`, so nothing gets written.

When reading the snapshot of the shard, e.g. during the first call to `PersistHandle::next` in `ReclockOperator::sync`, if you **do not** insert, in memory, `(Partitioned<PartitionId, MzOffset>::minimum(), Timestamp::minimum(), 1)`, then `ReclockOperator`'s `source_upper` is empty as is your `initial_batch` from `ReclockOperator::new`.

And lastly, the empty antichain is invalid to convert to `OffsetAntichain`, which is required for `mint_compat`, because the empty antichain is not domain-covering. This winds up failing before we have a chance to even look at the `source_frontier` in `mint_compat`.

The code then not only deals with the initialization, but also has to synthesize the retraction of `Partitioned<PartitionId, MzOffset>::minimum()`.

If you have any suggestions for a cleaner way of pretending that `Partitioned<PartitionId, MzOffset>::minimum()` is read, I'm very game.